zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -211,95 +211,123 @@ impl TransactionRecordsById {
-
-        let orchard_spends = transaction_record
+            .collect::<Result<Vec<&SaplingNote>, FeeError>>()
+    }
+    fn get_orchard_notes_spent_in_tx(
+        &self,
+        query: &TransactionRecord,
+    ) -> Result<Vec<&OrchardNote>, FeeError> {
+        query
             .spent_orchard_nullifiers()
             .iter()
             .map(|nullifier| {
                 self.values()
-                    .flat_map(|transaction_record| transaction_record.orchard_notes())
+                    .flat_map(|wallet_transaction_record| wallet_transaction_record.orchard_notes())
                     .find(|&note| {
                         if let Some(nf) = note.nullifier() {
                             nf == *nullifier
                         } else {
                             false
                         }
                     })
-                    .ok_or(FeeError::SpendNotFound)
+                    .ok_or(FeeError::OrchardSpendNotFound(*nullifier))
             })
-            .collect::<Result<Vec<&OrchardNote>, FeeError>>()?;
-        let orchard_spend_value: u64 = orchard_spends.iter().map(|&note| note.value()).sum();
-
-        let total_spend_value = transaction_record.total_transparent_value_spent
+            .collect::<Result<Vec<&OrchardNote>, FeeError>>()
+    }
+    /// Calculate the fee for a transaction in the wallet
+    ///
+    /// # Error
+    ///
+    /// Returns [`crate::wallet::error::FeeError::ReceivedTransaction`] if no spends or outgoing_tx_data were found
+    /// in the wallet for this transaction, indicating this transaction was not created by this spend capability.
+    /// Returns
+    /// [`crate::wallet::error::FeeError::SaplingSpendNotFound`]
+    /// OR
+    /// [`crate::wallet::error::FeeError::OrchardSpendNotFound`]
+    /// if any shielded spends in the transaction are not
+    /// found in the wallet, indicating that all shielded spends have not yet been synced.
+    /// Also returns this error
+    /// if the transaction record contains outgoing_tx_data but no spends are found.
+    /// If a transparent spend has not yet been synced, the fee will be incorrect and return
+    /// [`crate::wallet::error::FeeError::FeeUnderflow`] if an underflow occurs.
+    /// The tracking of transparent spends will be improved on the next internal wallet version.
+    pub fn calculate_transaction_fee(
+        &self,
+        query_record: &TransactionRecord,
+    ) -> Result<u64, FeeError> {
+        let sapling_spend_value: u64 = self
+            .get_sapling_notes_spent_in_tx(query_record)?
+            .iter()
+            .map(|&note| note.value())
+            .sum();
+        let orchard_spend_value: u64 = self
+            .get_orchard_notes_spent_in_tx(query_record)?
+            .iter()
+            .map(|&note| note.value())
+            .sum();
+        let total_trans_spent = query_record.total_transparent_value_spent;
```

Comment:
this line is unecessary it only used in the next line and has already caused a bug where you now have added query_record.total_transparent_value_spent twice.